### ISSUE TITLE
Profiling low-privileged processes with perf_events

### ISSUE THREAD

**Issue by Jongy** [jvm-profiling-tools/async-profiler#397]
Lately I was thinking how we can work around the issues with profiling low-privileged processes. The restrictions listed [here](https://github.com/jvm-profiling-tools/async-profiler#profiling-java-in-a-container) about containers, specifically, require a change in the container configuration, which means it has to be restarted (I'm referring to the seccomp limitations).
Additionally, system-wide settings have to be "weakened", security-wise: `perf_event_paranoid=1` and `kptr_restrict=0`, if we are to profile a non-root process.

Eventually it boils down to 2 things that require these modifications:
1. The `perf_event_open` call - blocked by seccomp in default Docker settings, and requires `perf_event_paranoid=1` or a privileged process to succeed. After creating the fd, there's no further need for privileges.
2. Reading from `/proc/kallsyms` - possibly requires changing `kptr_restrict` and/or having the relevant capabilities.

To tackle those issues seamlessly, here's what I think we can do:
1. Avoid calling `perf_event_open` in the target process. Make the call in another, high privileged process (the one invoking the profiler) and use UDS's `SCM_RIGHTS` to pass `perf_events` fds to the target process.
2. Avoid resolving kernel symbols in the process context. In general - I think it's better if as few work as possible is done in the context of the target process; resolving of Java symbols obviously belongs in the process, due to the use of AGCT; resolving of usermode symbols can be done "outside" but requires more effort). On the other hand, resolving of kernel symbols, processing the stack traces, emitting stackcollapse files/flamegraphs etc can be done by whomever invoked the profiler. I saw you already mentioned this possibility it https://github.com/jvm-profiling-tools/async-profiler/issues/353#issuecomment-702971124. If we stream back the raw samples, the high-privileged caller can resolve kernel symbols, and we avoid this issue completely.

The second point requires a big architectural change, so I wanted to raise it here and ask for your mind about it.

As for the first point, I have written a PoC which you can [find in my fork](https://github.com/Jongy/async-profiler/tree/perf-events-low-privs-poc). Basically, async-profiler no longer calls `perf_event_open` but instead creates an abstract-namespace UDS. After running `jattach ...`, I run another small binary `send_fds` which connects to async-profiler's UDS and handles "fd requests" - asnyc-profiler sends TIDs for which it wants the `perf_events` fd, and the high privileged `send_fds` creates them and sends them with `SCM_RIGHTS`.

This is all very messy right now:
* the protocol is very simple and doesn't support perf attributes, so I hardcoded the default  `cpu` event settings for `perf_event_open`
* I don't verify the requested TIDs are actually threads of the Java process
* `send_fds` itself should probably be incorporated into `jattach`? (problematic now because `jattach` drops privileges)
* I always use `send_fds` - I think it should be optional (and possibly, still call `perf_event_open` in async-profiler, if requested)
* mixed C/C++, code duplication, ...

But it works! :)

I run a simple, low-privs Java container:
```
$ docker run --user 8888:8888 --rm -it -v $(pwd)/test/Target.java:/tmp/Target.java -v$(pwd):$(pwd) -w /tmp java bash -c "javac Target.java && java Target"
```

Note that `/proc/sys/kernel/perf_event_paranoid` is `3`.

Run the profiler:
```
$ sudo ./profiler.sh -d 5 -e cpu -f /tmp/x.html -o flamegraph $(pgrep java)
```

and in another shell:
```
$ sudo nsenter -t $(pgrep java) -n -p build/send_fds $(grep NStgid /proc/$(pgrep java)/status | rev | cut -d$'\t' -f1 | rev)
```

This works. We don't get kernel stacks due to the 2nd point, tho. But I added some prints to see which addresses we get in the samples, and we clearly see kernel addresses:
```
...
ptr: 0xffffffff93aa819f
ptr: 0xffffffff93b91e7e
ptr: 0xffffffff93bc5932
ptr: 0xffffffff93bc69d2                           
ptr: 0xffffffff93b913b3
...
```

If we run the container w/o `--user ...` (so it remains UID 0) and grant it `CAP_SYSLOG`, it can successfully resolve kernel symbols, and we get a proper flamegraph with kernel functions.

What do you think, @apangin ? I could use some guidance here about the design, and about how this could be incorporated into async-profiler, if you think it's worthy. I'd be happy to continue working on it to get it merged, I think it's a useful addition to async-profiler.

**Comment by apangin** [comment]
Hi,
Thank you for sharing your thoughts.
I find the idea of passing perf_event descriptors from a privileged process quite interesting.

Recently I was thinking about a different problem, also related to profiling in containers. Unfortunately, the current approach with a perf_event per thread does not scale well across numerous containers on the same host. What I wanted to do is to establish a system-wide profiling process. It would collect kernel and native stacks, but in order to get Java stacks, it would need to communicate with the target process via some protocol. In particular, this would solve the problem of low-privileged containers from another side.

While the above is a long-term task, what I like in your approach is its simplicity - it is quite non-intrusive to the current profiler desing. Everything it needs is to substitute `perf_event_open` syscall.

I'm not sure if the mechanism you describe needs to be fully implemented in async-profiler, but at least I may agree with the minimal changes that will help to achieve the goal. I mean

- a new profiler option to pass the socket name for communication with a privileged agent;
- an alternative implementation of `perf_event_open` that will communicate over a socket instead of a syscall. This one will be used when the above option is specified.

A simple way to resolve kernel symbols is to provide another option for an alternative file name. When specified, profiler will read the kernel symbol map from the given file instead of `/proc/kallsyms`.

**Comment by Jongy** [comment]
> Recently I was thinking about a different problem, also related to profiling in containers. Unfortunately, the current approach with a perf_event per thread does not scale well across numerous containers on the same host. What I wanted to do is to establish a system-wide profiling process. It would collect kernel and native stacks, but in order to get Java stacks, it would need to communicate with the target process via some protocol. In particular, this would solve the problem of low-privileged containers from another side.

Yes, that would also solve the privileges problem. But, how will it work? As far as I understand async-profiler, you're using the `SIGPROF` emitted by perf_events to run code in usermode, in order to capture the exact state of Java frames (with AGCT) that'll match the kernel & user frames collected by perf_events. If we switch to a single, global perf_events instance - how will you accurately synchronize the frames collected by AGCT and perf_events? That sounds interesting...

About the mechanism I described: let's break it down to these components, then if it sounds good to you I'll try to find some time to develop it soon:
* a new option to `profiler.sh` that tells async-profiler to open a socket for comms instead of calling `perf_event_open` itself.
* that option will also invoke a standalone executable that'll connect to async-profiler and will create the perf_events fds as requested. Since it needs to run before jattach returns, it will be run asynchronously, started before jattach itself (and will attempt to connect to the target socket N times; async-profiler itself will `accept` a connection with some timeout, after which it'll give up and report failure)
* for the code of that binary ^^ I will reuse the `PerfEvents` class of async-profiler. The protocol between the two will pass `PerfEventType` structs so I can support all event types & settings.
* as for kernel symbols: I can allow passing an alternative file, but that'd require the user to take care of it. I think it's best if things are more automatic? So I can pass an open fd of `/proc/kallsyms`, as well... Looking in v5.12 sources, it should work - the "permissions" check happens during `open`, although I recall some kernel versions had it different. I'll check, and if it's not possible to pass an fd of `/proc/kallsyms` itself, I can have the new executable copy aside `/proc/kallsyms` then pass an fd of the copied version. This way, at any point we won't have a copy of `/proc/kallsyms` that's accessible to unprivileged users.

So eventually you'll use it as `./profler.sh --unpriv-perf` and it will "just work" (parameter naming's debatable :)

**Comment by apangin** [comment]
> how will you accurately synchronize the frames collected by AGCT and perf_events

ePBF program attached to the perf_event descriptor can collect kernel stack trace and send profiling signal to the application whenever perf_event fires. I haven't tried this yet though.

> The protocol between the two will pass PerfEventType structs so I can support all event types & settings.

I'd suggest to pass `struct perf_event_attr` instead. This will be simple and universal. The standalone program won't need to share the code or be in sync with async-profiler's internals. For example, I've just pushed the change that adds support for custom PMU events and kprobes; this modifies `PerfEventType` structure, but this would just work with an alternative `perf_event_open` implementation.

> I can allow passing an alternative file, but that'd require the user to take care of it. I think it's best if things are more automatic? So I can pass an open fd of /proc/kallsyms, as well.

Why passing an alternative file can't be automated? IMO, it's again simpler and more universal approach. This does not even need to be a regular file - for example, one could pass a named pipe, and stream the contents of `kallsyms` through it.

If there is a socket for passing perf_events anyway, the contents of `kallsyms` could be as well transferred during the handshake. But I like the idea of an alternative filename, as it requires minimum changes to the async-profiler core, and it can be useful in some environments even without an external program.

**Comment by Jongy** [comment]
> ePBF program attached to the perf_event descriptor can collect kernel stack trace and send profiling signal to the application whenever perf_event fires. I haven't tried this yet though.

I see. I thought you intend to avoid the AGCT itself as well (that is, a fully external profiler). But indeed an eBPF program that signals the proper thread can also be made system-wide. I'm just not sure that eBPF has the proper API for that ([bpf_send_signal](https://man7.org/linux/man-pages/man7/bpf-helpers.7.html) sends the signal to any of the threads, and not necessarily to the thread which runs the eBPF program)

> I'd suggest to pass `struct perf_event_attr` instead. This will be simple and universal. The standalone program won't need to share the code or be in sync with async-profiler's internals. For example, I've just pushed the change that adds support for custom PMU events and kprobes; this modifies `PerfEventType` structure, but this would just work with an alternative `perf_event_open` implementation.

Haha, that's a great idea, `perf_event_attr` already exists for the "RPC" between the kernel & userspace, I'll just reuse that same structure instead of `PerfEventType`.
 
> Why passing an alternative file can't be automated? IMO, it's again simpler and more universal approach. This does not even need to be a regular file - for example, one could pass a named pipe, and stream the contents of `kallsyms` through it.

It can - sorry if I was misunderstood, the `I think it's best if things are more automatic` part was referring to the current state, not to the new suggestion. I raised it as a question to consider doing - obviously it can be done with rather ease.

> If there is a socket for passing perf_events anyway, the contents of `kallsyms` could be as well transferred during the handshake. But I like the idea of an alternative filename, as it requires minimum changes to the async-profiler core, and it can be useful in some environments even without an external program.

Eventually I went with passing a file descriptor. Mostly for "security" reasons - this way, there's no need to have any copy of `/proc/kallsyms` dropped on the filesystem, available for low-privileged users. It did make the core of async-profiler a bit more complicated - we can discuss that in the PR and possibly go with a simple file.

I've finally found some time to work on it yesterday, I have a working branch with the changes as we have discussed in this issue. I will do some cleanups and open a PR.

**Comment by apangin** [comment]
Resolved by #411
